crates/sound/src/device.rs

```diff
@@ -388,42 +388,51 @@ impl VhostUserSoundThread {
             return Ok(true);
         }
 
+        // Instead of counting descriptor chain lengths, encode the "parsing" logic in
+        // an enumeration. Then, the compiler will complain about any unhandled
+        // match {} cases if any part of the code is changed. This makes invalid
+        // states unrepresentable in the source code.
         #[derive(Copy, Clone, PartialEq, Debug)]
-        enum TxState {
+        enum IoState {
             Ready,
             WaitingBufferForStreamId(u32),
             Done,
         }
 
+        // Keep log of stream IDs to wake up, in case the guest has queued more than
+        // one.
         let mut stream_ids = BTreeSet::default();
 
         for desc_chain in requests {
-            let mut state = TxState::Ready;
+            let mut state = IoState::Ready;
             let mut buffers = vec![];
             let descriptors: Vec<_> = desc_chain.clone().collect();
             let message = Arc::new(IOMessage {
                 vring: vring.clone(),
                 status: VIRTIO_SND_S_OK.into(),
                 latency_bytes: 0.into(),
                 desc_chain: desc_chain.clone(),
-                descriptor: descriptors.last().cloned().unwrap(),
+                response_descriptor: descriptors
+                    .last()
+                    .cloned()
+                    .expect("Received IO request with an empty descriptor chain."),
```

Comment:
Should we return an error to the guest instead of panicking?

Reply:
Oh yes, good point.